rosbag2_transport/src/rosbag2_transport/rosbag2_node.cpp

```diff
@@ -149,24 +151,44 @@ std::unordered_map<std::string, std::string> Rosbag2Node::get_topics_with_types(
 }
 
 std::unordered_map<std::string, std::string>
-Rosbag2Node::get_all_topics_with_types()
+Rosbag2Node::get_all_topics_with_types(bool include_hidden_topics)
 {
-  return filter_topics_with_more_than_one_type(this->get_topic_names_and_types());
+  return filter_topics_with_more_than_one_type(
+    this->get_topic_names_and_types(), include_hidden_topics);
 }
 
 std::unordered_map<std::string, std::string> Rosbag2Node::filter_topics_with_more_than_one_type(
-  std::map<std::string, std::vector<std::string>> topics_and_types)
+  const std::map<std::string, std::vector<std::string>> & topics_and_types,
+  bool include_hidden_topics)
 {
   std::unordered_map<std::string, std::string> filtered_topics_and_types;
   for (const auto & topic_and_type : topics_and_types) {
     if (topic_and_type.second.size() > 1) {
       ROSBAG2_TRANSPORT_LOG_ERROR_STREAM(
         "Topic '" << topic_and_type.first <<
           "' has several types associated. Only topics with one type are supported");
-    } else {
-      filtered_topics_and_types.insert({topic_and_type.first, topic_and_type.second[0]});
+      continue;
+    }
+
+    // According to rclpy's implementation, the indicator for a hidden topic is a leading '_'
+    // https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/topic_or_service_is_hidden.py#L15
+    if (!include_hidden_topics) {
+      auto tokens = rcpputils::split(topic_and_type.first, '/', true);  // skip empty
+      auto is_hidden = std::find_if(tokens.begin(), tokens.end(), [](const auto & token) -> bool {
+          return token[0] == '_';
+          });
+      if (is_hidden != tokens.end()) {
+        ROSBAG2_TRANSPORT_LOG_ERROR_STREAM(
+          "Hidden topics are not recorded. Topic '" <<
+            topic_and_type.first << "' will be skipped." <<
+            "You can enable them with --include-hidden-topics");
```

Comment:
Since this is polled periodically, I am looking for a way to log only once. I am currently unable to find the appropriate macro, any ideas?